cmd/blockhashes.cpp

```diff
@@ -89,7 +89,7 @@ int main(int argc, char* argv[]) {
             collector.load(blockhashes_table.get(), nullptr, 0);
         }
         // Update progress
-        db::stages::set_stage_progress(*txn, db::stages::kBlockHashesKey, current_block_number);
+        db::stages::set_stage_progress(*txn, db::stages::kBlockHashesKey, current_block_number-1);
```

Comment:
This is still not logically correct.
Picture this example where stage height is 10
```
        auto current_block_number{db::stages::get_stage_progress(*txn, db::stages::kBlockHashesKey)};
        auto initial_block_number = current_block_number;
```
After this lines current_block_number == initial_block_number == 10.
Here we have the first problem as the stage has already processed block 10 while the loop should begin from block 11
Then we jump to the scan loop
```
        for (int rc{header_table->seek(&key_mdb, &data_mdb)}; rc != MDB_NOTFOUND;
             rc = header_table->get_next(&key_mdb, &data_mdb)) {
            // Check if it's an header entry
            if (key_mdb.mv_size != 40) continue;
            Bytes key(static_cast<unsigned char*>(key_mdb.mv_data), key_mdb.mv_size);
            // We set the key to the hash of the header and the value to the block number
            etl::Entry entry{key.substr(8, 40), key.substr(0, 8)};
            collector.collect(entry);
            current_block_number++;
        }
```
If you look at the loop it will exit only if `rc == MDB_NOT_FOUND` which means any IO or txn error will make you loop indefinitely.
But imagine on first loop we get an MDB_NOT_FOUND: you still have current_block == 10 ... and you will end up
setting stage progress to
```
db::stages::set_stage_progress(*txn, db::stages::kBlockHashesKey, current_block_number - 1);
```
which will end up with stage set to 9.

Reply:
so, just setting it to current_block_number as i did before was correct, because it would end up in process the correct amount of blocks and resume from so that the correct blocks were processed, no?

Reply:
Not really ... as I showed in the discord channel with your previous revision if you had processed last block as 10 you'd end up setting stage height at 11

Reply:
you process block 10 previously and now you are about to process 11, i don't see whats wrong with it. you finished 10 in the previous cycle and you do not repeat it again.

Reply:
Stage height is not the block you're about to process: is the last block processed.

Reply:
ok now it should as yo said